### PR TITLE
Run the npgsql/npgsql tests in our pipelines with the profiler attached

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Changes proposed in this pull request:
- Adds a new Azure DevOps pipeline to run integration tests from libraries, starting with Npgsql
  - Runs Npgsql tests at tags v4.1.9 and v5.0.7, on both Ubuntu and Windows, with automatic instrumentation enabled
  - It re-uses the build steps from the consolidated pipeline to build the .NET Tracer artifacts
  - Scheduled to run daily at 3am UTC
  - Sets up Postgres with several extensions like PostGIS. These setup instructions were copied directly from the source repo

Results
- Several tests fail on the v5.0.7 tag. The root cause appears to be the runtime types of exception objects being different when Datadog automatic instrumentation is attached AND CallTarget is enabled
  - Npgsql.Tests.CommandTests.CancelAsyncHard
  - Npgsql.Tests.SecurityTests.Bug1718
  - Npgsql.Tests.CommandTests.CancelAsyncSoft
  - Npgsql.Tests.CommandTests.CancelAsyncImmediately

@DataDog/apm-dotnet

### CODE REVIEW

File: .azure-pipelines/third-party-test-suites.yml
Comment (andrewlock) [inline .azure-pipelines/third-party-test-suites.yml:361]: Could probably make these a Nuke Target, but given they're not doing much, probably easier not to bother!

File: .azure-pipelines/third-party-test-suites.yml
Comment (andrewlock) [inline .azure-pipelines/third-party-test-suites.yml:193]: Should we add Alpine + Arm64 to this matrix?

File: .azure-pipelines/third-party-test-suites.yml
Comment (andrewlock) [inline .azure-pipelines/third-party-test-suites.yml:352]: I think `4.1.9` should be `$(tagName)`? And maybe we want to split by os?
```suggestion
        dd-trace --dd-env CI --dd-service npgsql.tests.$(os) --dd-version $(tagName) dotnet "test npgsql/test/Npgsql.Tests --logger trx;LogFileName=npgsql.tests.$(tagName).trx --results-directory npgsql/test/results"
```

**Review by andrewlock — APPROVED** [review]
LGTM, I don't see the test failures though, are we explicitly ignoring them?

Also, there's a lot (200 lines) of duplication with the ultimate-pipeline. I wonder if we should either:

* Move the final stage to the ultimate-pipeline, but only run it on the scheduled build
* Extract all the build stages to be "steps" that can be imported in both places (doesn't remove the duplication in variables etc)
* Update this pipeline to pull the dd-trace directly from the latest master build. This would let you remove all the build and tool stages, and just pull the latest dd-tool from the latest master branch build.

The final option seems like quite an elegant solution to me.

File: .azure-pipelines/third-party-test-suites.yml
Comment (tonyredondo) [inline .azure-pipelines/third-party-test-suites.yml:358]: Non blocking, just a personal preference...

I think you can avoid the `"` by using a `--` before the final command to wrap.
Also I prefer to use `=` to make all option values more clear.

```suggestion
        dd-trace --dd-env=CI --dd-service=npgsql.plugintests --dd-version=v4.1.9 -- dotnet test npgsql/test/Npgsql.PluginTests --logger trx;LogFileName=npgsql.plugintests.$(os).$(tagName).trx --results-directory npgsql/test/results
```

**Comment by zacharycmontoya** [comment]
> Update this pipeline to pull the dd-trace directly from the latest master build. This would let you remove all the build and tool stages, and just pull the latest dd-tool from the latest master branch build.

This is a great idea. I was able to get implement it in 8187315 . Since some of our master builds don't build the dotnet tool, I have to add a build tag to the conslidated-pipeline to specify which of our builds produce it, and then the new pipeline will search builds with the tag to find the latest artifact to download.

File: .azure-pipelines/third-party-test-suites.yml
Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml]: Yeah that's what I was thinking 😄 Maybe if some other repo testing gets to be more complicated, we could reconsider

File: .azure-pipelines/third-party-test-suites.yml
Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml]: Thanks for the suggestion, I applied the `option=value`.
For the `"` style, I tried to remove them but the command line was interpreting `-- results-directory` as a separate command after running the `dd-trace` command, so I put them back

File: .azure-pipelines/third-party-test-suites.yml
Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml:49]: I wasn't considering adding those because I was focused on running only the platforms that the repo was already testing on. This way I don't spend a lot of time trying to figure out how to do the setup on new platforms. Do you think the current Windows/Ubuntu testing is sufficient?

File: .azure-pipelines/third-party-test-suites.yml
Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml]: Good catch on the tag name. For the operating system, our CI already adds a tag to the tests by the OS Platform so I'm thinking of not adding the OS to the service name. Any objections @andrewlock @tonyredondo ?

**Comment by zacharycmontoya** [comment]
> LGTM, I don't see the test failures though, are we explicitly ignoring them?
Since I had to publish the test results separately via the `PublishTestResults@2` task, the script to run the tests exited successfully and the test results upload didn't automatically trigger a failure. I kept those defaults just to make sure that I could tell things were working while I continued to tweak the run, but now I'm turning on the setting `failTaskOnFailedTests: true` so that failed tests will cause the build to fail.

**Comment by zacharycmontoya** [comment]
At this time, I think I've addressed all of the feedback and you can see from the testing of commit 88e4dea8cf47da57004f733db3ffcb60ad4f01af, the Third Party Test Suite is failing now with those four tests I listed, which is fine for now. Once I get approvals (so I don't need to run the test suite pipeline again) I'll modify the test run to grab the latest dotnet-tool build from master instead of this branch.

File: .azure-pipelines/third-party-test-suites.yml
Comment (andrewlock) [reply .azure-pipelines/third-party-test-suites.yml:50]: Ah, good point. I was thinking about what _we_ run on, but yeah agreed, we definitely don't want to take on that complexity

File: .azure-pipelines/third-party-test-suites.yml
Comment (lucaspimentel) [inline .azure-pipelines/third-party-test-suites.yml:20]: Aren't these moving to `$(System.DefaultWorkingDirectory)/bin/...` (dropping the `src`)

File: .azure-pipelines/third-party-test-suites.yml
Comment (lucaspimentel) [inline .azure-pipelines/third-party-test-suites.yml:88]: Is this a TODO for later?

Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml:88]: Yes, before merging to master I'll swap out the lines 😄 Until I merge this, I can't pull build artifacts from master, only this branch which has the required change.

**Review by lucaspimentel — APPROVED** [review]

File: .azure-pipelines/third-party-test-suites.yml
Comment (zacharycmontoya) [reply .azure-pipelines/third-party-test-suites.yml:20]: I copied and pasted the values from our consolidated pipeline yml. I think you're right that the location has changed, but maybe we're no longer using some of these variables

**Comment by zacharycmontoya** [comment]
Getting ready to merge so I updated the `runBranch` to grab Tracer artifacts from master in the latest commit

File: .azure-pipelines/third-party-test-suites.yml
Comment (andrewlock) [reply .azure-pipelines/third-party-test-suites.yml:20]: We should probably update the consolidated pipeline, but it's not a big issue. These variables override the "default" values, which, as you say, has dropped the src.

**Pull request merged**
